### ISSUE TITLE
[Feature] Exchange Mailboxes

### ISSUE THREAD

**Issue by grantholle** [DirectoryTree/LdapRecord#171]
I'm about to tackle mailbox creation in a new codebase. In an old one I'm using the [ancient] [LdapTools](https://github.com/ldaptools/ldaptools) package to handle creating a user and mailbox in a single creation request. I looks like it's just adding the `msExchMailboxGUID` filter to the user objectClass, but this is an area I'm not too familiar ([details](https://github.com/ldaptools/ldaptools/blob/master/docs/en/reference/Default-Schema-Attributes.md#exchange-mailbox-user-types)). Since it's still just a "user", wouldn't just adding the right attributes create the mailbox automatically?

I would love to submit a PR for this after some tinkering.

**Comment by grantholle** [comment]
Also, would you be opposed to adding all the different Exchange entity/model types?

**Comment by stevebauman** [comment]
Hi @grantholle,

I'm not experienced in this front -- but I'm absolutely all for patching this in and having the functionality available. I'd be happy to accept a PR!

> Also, would you be opposed to adding all the different Exchange entity/model types?

Definitely not. The more features available, the better. If that means adding models for each exchange entity, that's totally fine with me 👍 .

Thanks for taking the time to investigate this!

**Comment by stevebauman** [comment]
[v1.9.0](https://github.com/DirectoryTree/LdapRecord/releases/tag/v1.9.0) has just been released with the `User::whereHasMailbox()` scope 👍

**Comment by grantholle** [comment]
I'm working on trying to pull my Exchange databases. I can retrieve them using LdapTools, but can't seem to get them unless I change my base dn. The catch is that I don't know how to know which dn to change it to.

LdapTools is really abstracted so it's hard to follow exactly what it's doing. I can't find where it's running the preliminary query to get the base dn. Here is the logging output for LdapTools running a query to get my [Exchange Database](https://github.com/ldaptools/ldaptools/blob/master/docs/en/reference/Default-Schema-Attributes.md#exchange-database-types).

```
(myserver.com) -- Start Operation Type: Query
	Filter: (&(objectClass=*))
	Base DN: 
	Attributes: configurationNamingContext,defaultNamingContext,schemaNamingContext,supportedControl,namingContexts,rootDomainNamingContext,supportedSaslMechanisms,supportedLdapPolicies,supportedLdapVersion,vendorName,vendorVersion,isSynchronized,isGlobalCatalogReady,domainFunctionality,forestFunctionality,domainControllerFunctionality,domainFunctionality,forestFunctionality,domainControllerFunctionality,dsServiceName,currentTime
	Scope: base
	Use Paging: false
	Page Size: 1000
	Size Limit: 0
	Use Cache: false
	Execute on Cache Miss: true
	Invalidate Cache: false
	Server: myserver.com
	Controls: array (
)
	Cache Hit: false
(myserver.com) -- End Operation Type: Query -- (0.019635915756226 seconds)
(myserver.com) -- Start Operation Type: Query
	Filter: (&(&(objectClass=msExchMDB))(&(cn=16-TianJinSiRui)))
	Base DN: CN=Configuration,DC=LDi,DC=LDiChina,DC=com
	Attributes: cn,distinguishedName,objectGuid
	Scope: subtree
	Use Paging: true
	Page Size: 1000
	Size Limit: 0
	Use Cache: false
	Execute on Cache Miss: true
	Invalidate Cache: false
	Server: myserver.com
	Controls: array (
)
	Cache Hit: false
(myserver.com) -- End Operation Type: Query -- (0.02168083190918 seconds)
```

I can get my database by doing the following query (now that I know the correct base dn):

```php
$connection->query()
    ->in('CN=Configuration,DC=LDi,DC=LDiChina,DC=com')
    ->where('objectclass', 'msExchMDB')
    ->get();
```

However, without providing an `in()`, it doesn't return any results. From looking at the logging, do you know what that first query is and its purpose? It's obviously finding some dn that contains the info I was looking for when trying to get my Exchange server and database info.

**Comment by grantholle** [comment]
Ok, so I've figured out it's querying the "rootdse". I'm not sure how to do that using LdapRecord. Perhaps it has something with the scope being `base`? Setting `->in('')` doesn't return any results. I'm totally out of my element at this point.

**Comment by stevebauman** [comment]
Hey @grantholle, to query the Root DSE of your Active Directory server, you must switch the query to a `read` operation instead of a `search` or `listing`:

```php
// Performing a raw connection query:
$rootDse = $connection->query()
    ->in(null)
    ->read()
    ->whereHas('objectclass')
    ->first();

// Using the base `Entry` model:

use LdapRecord\Models\Entry;

$rootDse = Entry::in(null)
            ->read()
            ->whereHas('objectclass')
            ->first();
```

That should work for you!

**Comment by stevebauman** [comment]
I'll create an issue to add this into the documentation 👍

**Comment by grantholle** [comment]
You're awesome @stevebauman, thanks 💯

**Comment by stevebauman** [comment]
Thanks for your kind words @grantholle! :smile:
